NIFI-738: Improve error handling for CSV and JSON conversion.

rdblue opened the pull request:
Changes:
* Send bad record information on the "incompatible" relationship
* Use an attribute, "errors" for summarized error messages
* Send no error content for now, original content can't be accessed
* Summarize similar error messages with "N similar failures"
* Add similar error handling to CSV conversion

joewitt commented:
Builds cleanly including the contribution check for licenses and check style.

ConvertCSVToAvro
- Line 258 should use parameters passed in via object[] rather than string concatenation.

- What happens if you have processed an entire input and produced no successful results due to a DatasetRecordExceptions?  I think it would forward an empty flow file since nothing would have been written to it.  Is that what you want?  Such a case seems like a FAILURE to me.

- Consider removing the incompatible relationship and simply adding an attribute to the success flow file which lists the reasons or add an attribute per reason for difficulty converting the original.  If you will keep the incompatible relationship then I’d recommend sending it a clone of the original item and then adding the relationship.  That would make troubleshooting or analysis of that object much easier.  In any case be very careful about how large these attributes could become.  If you’re storing a series of stack traces this could become unwieldy and memory inefficient really fast.

ConvertJSONToAvro
- Line 160 should use parameters passed in via object[] rather than string concatenation.

- Same comments as Convert CSV to Avro.

FailureTracker
- This class is a cool idea.  My concern with it is that it appears to store all the raw reasons which could be quite significant in large files with *many* thousands of lines.  This could create issues with the heap, cause excessive GC, etc.  Consider refactoring this class to both immediately store the summarized representation and to bound the total number of reasons it will store.  It can just ignore additional errors as they arrive.  This will allow it to behave well in terms of memory consumption even in the presence of massive failures as everything will be bounded.

Thanks
Joe

rdblue commented:
Thanks for taking a look, Joe.

I implemented the changes as you suggested. I kept the incompatible relationship because when we can recover the original content, I intend to filter out the successful records and send just the failures. But you're right that for now we should include the original data to make it easier to debug flows.

For the failure tracker, I updated it to keep just the first full error message and a counter instead of the multimap. I don't think it will be a memory problem now.

joewitt commented:
Ryan

Please add the below unit test to TestCVSToAvroProcessor.  It will fail and explains why nicely.  This issue is present in AvroToCSV as well I believe.

Only other thing I noticed was that you did not cap the number of errors/reasons you'll store.  I just bring this up as a 'be sure this is fine' thing.  It isn't clear if this is important in a practical sense.

```java
    @Test
    public void testBasicConversionNoErrors() throws IOException {
        TestRunner runner = TestRunners.newTestRunner(ConvertCSVToAvro.class);
        runner.assertNotValid();
        runner.setProperty(ConvertCSVToAvro.SCHEMA, SCHEMA.toString());
        runner.assertValid();

        runner.enqueue(streamFor("1,green\n2,blue,\n3,grey,12.95"));
        runner.run();

        long converted = runner.getCounterValue("Converted records");
        long errors = runner.getCounterValue("Conversion errors");
        Assert.assertEquals("Should convert 2 rows", 2, converted);
        Assert.assertEquals("Should reject 0 row", 0, errors);

        runner.assertTransferCount("success", 1);
        runner.assertTransferCount("failure", 0);
        runner.assertTransferCount("incompatible", 0);

    }
```

rdblue commented:
@joewitt, I get an error that it converted 3 rows successfully instead of 2. I'm not sure I understand why that isn't the correct behavior. Shouldn't each CSV row be converted?

rdblue commented:
Nevermind, I was in the wrong branch. I've fixed the bug and will have this updated shortly. Thanks!

joewitt commented:
Cool.  So you think this is ready to be merged now then?  Also, just want to be sure you don't think there is any practical issue with leaving the number of errors it will track unbounded.

rdblue commented:
I think this is ready to go. I'm not concerned with the error tracking because we keep to a pretty reasonable format, "General error: specific data". The general errors will be stored along with a count and a single full error message. We don't have very many possibilities for that, usually "Can't convert to <type>" or "Field <name>", both of which are pretty limited.

joewitt commented:
Roger.  I'm a +1 now.  Will merge in a bit.  Thanks for addressing it all so quickly.

rdblue commented:
No problem, I wanted to get it into the release. Thanks for your help!

joewitt commented:
```
ests run: 4, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0.339 sec <<< FAILURE! - in org.apache.nifi.processors.kite.TestJSONToAvroProcessor
testBasicConversion(org.apache.nifi.processors.kite.TestJSONToAvroProcessor)  Time elapsed: 0.325 sec  <<< FAILURE!
org.junit.ComparisonFailure: Should accumulate error messages expected:<...annot convert field [id: Cannot convert to long: "120V", Cannot convert field color: Cannot convert to string: 15.23 (1 similar failure)]> but was:<...annot convert field [color: Cannot convert to string: 15.23 (1 similar failure), Cannot convert field id: Cannot convert to long: "120V"]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.apache.nifi.processors.kite.TestJSONToAvroProcessor.testBasicConversion(TestJSONToAvroProcessor.java:83)

testOnlyErrors(org.apache.nifi.processors.kite.TestJSONToAvroProcessor)  Time elapsed: 0.004 sec  <<< FAILURE!
org.junit.ComparisonFailure: Should set an error message expected:<...annot convert field [id: Cannot convert to long: "120V", Cannot convert field color: Cannot convert to string: 15.23 (1 similar failure)]> but was:<...annot convert field [color: Cannot convert to string: 15.23 (1 similar failure), Cannot convert field id: Cannot convert to long: "120V"]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.apache.nifi.processors.kite.TestJSONToAvroProcessor.testOnlyErrors(TestJSONToAvroProcessor.java:107)

Running org.apache.nifi.processors.kite.TestKiteProcessorsCluster
Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0 sec - in org.apache.nifi.processors.kite.TestKiteProcessorsCluster
Running org.apache.nifi.processors.kite.TestKiteStorageProcessor
Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0 sec - in org.apache.nifi.processors.kite.TestKiteStorageProcessor

Results :

Failed tests:
  TestJSONToAvroProcessor.testBasicConversion:83 Should accumulate error messages expected:<...annot convert field [id: Cannot convert to long: "120V", Cannot convert field color: Cannot convert to string: 15.23 (1 similar failure)]> but was:<...annot convert field [color: Cannot convert to string: 15.23 (1 similar failure), Cannot convert field id: Cannot convert to long: "120V"]>
  TestJSONToAvroProcessor.testOnlyErrors:107 Should set an error message expected:<...annot convert field [id: Cannot convert to long: "120V", Cannot convert field color: Cannot convert to string: 15.23 (1 similar failure)]> but was:<...annot convert field [color: Cannot convert to string: 15.23 (1 similar failure), Cannot convert field id: Cannot convert to long: "120V"]>
```

Ryan i had trouble running the unit tests.  It is possible I've now returned the wrong branch favor but I think I'm building against exactly what you sent.  Can you confirm?

rdblue commented:
@joewitt, looks like the problem is that the accumulated error messages are out of order. This is a result of moving from the multimap to a plain hashmap. I'll update to use a linked hashmap, which will give a reliable order.
